internal/controller/git/work_tree.go

```diff
@@ -0,0 +1,514 @@
+	DeleteBranch(branch string) error
+	// Dir returns an absolute path to the working tree.
+	Dir() string
+	// HasDiffs returns a bool indicating whether the working tree currently
+	// contains any differences from what's already at the head of the current
+	// branch.
+	HasDiffs() (bool, error)
+	// HomeDir returns an absolute path to the home directory of the system user
+	// who cloned the repo associated with this working tree.
+	HomeDir() string
+	// GetDiffPathsForCommitID returns a string slice indicating the paths,
+	// relative to the root of the repository, of any files that are new or
+	// modified in the commit with the given ID.
+	GetDiffPathsForCommitID(commitID string) ([]string, error)
+	// IsAncestor returns true if parent branch is an ancestor of child
+	IsAncestor(parent string, child string) (bool, error)
+	// LastCommitID returns the ID (sha) of the most recent commit to the current
+	// branch.
+	LastCommitID() (string, error)
+	// ListTags returns a slice of tags in the repository with metadata such as
+	// commit ID, creator date, and subject.
+	ListTags() ([]TagMetadata, error)
+	// ListCommits returns a slice of commits in the current branch with
+	// metadata such as commit ID, commit date, and subject.
+	ListCommits(limit, skip uint) ([]CommitMetadata, error)
+	// CommitMessage returns the text of the most recent commit message associated
+	// with the specified commit ID.
+	CommitMessage(id string) (string, error)
+	// Push pushes from the current branch to a remote branch by the same name.
+	Push(force bool) error
+	// RefsHaveDiffs returns whether there is a diff between two commits/branches
+	RefsHaveDiffs(commit1 string, commit2 string) (bool, error)
+	// RemoteBranchExists returns a bool indicating if the specified branch exists
+	// in the remote repository.
+	RemoteBranchExists(branch string) (bool, error)
+	// ResetHard performs a hard reset on the working tree.
+	ResetHard() error
+	// URL returns the remote URL of the repository.
+	URL() string
+}
+
+// workTree is an implementation of the WorkTree interface for interacting with
+// any working tree of a Git repository.
+type workTree struct {
+	*baseRepo
+	bareRepo *bareRepo
+}
+
+type LoadWorkTreeOptions struct {
+	Credentials           *RepoCredentials
+	InsecureSkipTLSVerify bool
+}
+
+func LoadWorkTree(path string, opts *LoadWorkTreeOptions) (WorkTree, error) {
+	if opts == nil {
+		opts = &LoadWorkTreeOptions{}
+	}
+	w := &workTree{
+		baseRepo: &baseRepo{
+			creds:                 opts.Credentials,
+			dir:                   path,
+			insecureSkipTLSVerify: opts.InsecureSkipTLSVerify,
+		},
+	}
+	res, err := libExec.Exec(w.buildGitCommand("rev-parse", "--git-dir"))
+	if err != nil {
+		return nil, fmt.Errorf("error getting git directory: %w", err)
+	}
+	// The root of the bare repo will be three levels up from the path returned
+	// by git rev-parse --git-dir
+	repoPath := filepath.Join(strings.TrimSpace(string(res)), "..", "..", "..", "repo")
```

Comment:
Non-blocking: this feels prone to errors.

Reply:
I didn't love this either and I'd be happy to take any suggestion your might have.

Reply:
Refactored this in https://github.com/akuity/kargo/pull/2483/commits/b4a47b07829d6b326110b69d352a4bb134a4c5b9